### PR TITLE
[LAKESIDE-22] Update search so that category entities count is displayed correctly.

### PR DESCRIPTION
**Pull request by drtwisted**
**Description:** 
Counter beside the category name now shows actual results count for current search state s

**Youtrack:** https://youtrack.raccoongang.com/issue/FLS-22

### CODE REVIEW

File: common/djangoapps/course_category/search_engine.py [hunk 1]
```diff
@@ -15,53 +19,118 @@ def search(self, **kwargs):
...
+
+    Returns:
+        dict: A dictionary of programs data with extra `Without group` program
+              in case `courses` contains courses that does not belong to any program.
+    """
+    courses_ids = [course.get('data').get('id') for course in courses]
+    ungrouped_courses_ids = set(courses_ids) - already_grouped_courses_ids
+
+    if ungrouped_courses_ids:
+        ungrouped_courses = list(filter(lambda c: c.get('data', {}).get('id') in ungrouped_courses_ids, courses))
```
[the remark this answers is not in the document]
Comment (cmltaWt0) [reply common/djangoapps/course_category/search_engine.py]: I prefer to use a list comprehension:
```python
[_ for _ in courses if _.get('data', {}).get('id') in ungrouped_courses_ids]
```

[the remark this answers is not in the document]
Comment (drtwisted) [reply common/djangoapps/course_category/search_engine.py]: Yeah, it was for future python3 and to shorten up the line, but the comprehension is a preferred way, agreed.

File: common/djangoapps/course_category/search_engine.py [hunk 2]
```diff
@@ -74,6 +143,22 @@ def _search_programs(**kwargs):
 
     result_intersection = title_qs & category_qs
 
-    result = [{'title':p.title, 'subtitle':p.subtitle, 'courses':p.courses} for p in result_intersection]
+    return result_intersection
+
+
+def _format_programs(programs):
+    """
+    Format programs list accordingly to the scheme:
```
[the remark this answers is not in the document]
Comment (drtwisted) [reply common/djangoapps/course_category/search_engine.py]: Fixed.

File: common/djangoapps/course_category/search_engine.py [hunk 3]
```diff
@@ -15,53 +19,118 @@ def search(self, **kwargs):
...
+
+    return programs
+
+
 def _search_programs(**kwargs):
+    query_string = kwargs.get('query_string', '')
 
-    title = kwargs.get('query_string', '')
+    if query_string:
+        title_qs = Program.objects.filter(Q(title__icontains=query_string) | Q(subtitle__icontains=query_string))
```
[the remark this answers is not in the document]
Comment (drtwisted) [reply common/djangoapps/course_category/search_engine.py]: Improved function code.

**Pull request merged**